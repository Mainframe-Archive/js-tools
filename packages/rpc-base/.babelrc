{
  "presets": [
    [
      "env",
      {
        "targets": {
          "browsers": ["last 2 versions", "ie >= 11"],
          "node": 8
        },
        "modules": false,
        "loose": true
      }
    ],
    "flow"
  ],
  "plugins": ["syntax-class-properties", "transform-class-properties"],
  "env": {
    "test": {
      "presets": ["env", "flow"]
    }
  }
}
