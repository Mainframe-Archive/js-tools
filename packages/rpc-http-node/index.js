require = require('esm')(module, { mode: 'auto' })
module.exports = require('./lib/index.js')
