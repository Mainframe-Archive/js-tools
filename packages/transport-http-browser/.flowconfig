/Users/paul/dev/mainframe/js-tools/.flowconfig