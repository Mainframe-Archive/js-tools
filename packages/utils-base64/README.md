# utils-base64

Utilities for base64-encoded strings.

## Installation

```sh
yarn add @mainframe/utils-base64
```

## Usage

```js
import { encodeBase64, decodeBase64 } from 'utils-base64'

const buffer = Buffer.from('test')
const encoded = encodeBase64(buffer) // base64 (string)
const decoded = decodeBase64(encoded) // Buffer
```

## Types

### base64

Opaque type representing a base64-encoded string.

## API

### base64Type()

Simple identity function casting the provided value as `base64` to properly type-check.

**Arguments**

1.  `value: any`

**Returns** `base64`

### encodeBase64()

**Arguments**

1.  `input: Buffer`

**Returns** `base64`

### decodeBase64()

**Arguments**

1.  `input: base64`

**Returns** `Buffer`

## License

MIT
