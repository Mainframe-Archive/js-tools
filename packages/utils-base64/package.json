{
  "name": "@mainframe/utils-base64",
  "version": "0.1.0",
  "description": "Base64 encoding utilities",
  "repository": "git@github.com:MainframeHQ/js-tools.git",
  "main": "lib/index.js",
  "author": "Mainframe",
  "license": "MIT",
  "files": [
    "lib/*"
  ],
  "scripts": {
    "clean": "del lib",
    "build:js": "flow-remove-types src --out-dir lib",
    "build:flow": "flow-copy-source src lib",
    "build": "yarn clean && yarn build:js && yarn build:flow",
    "test:types": "flow check",
    "test:unit": "jest",
    "prepublishOnly": "yarn build"
  },
  "devDependencies": {
    "flow-bin": "^0.74.0"
  }
}
