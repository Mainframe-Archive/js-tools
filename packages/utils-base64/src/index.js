// @flow

export opaque type base64: string = string

export const base64Type = (value: any): base64 => (value: base64)

export const encodeBase64 = (input: Buffer): base64 => {
  return base64Type(input.toString('base64'))
}

export const decodeBase64 = (input: base64): Buffer => {
  return Buffer.from(input, 'base64')
}
