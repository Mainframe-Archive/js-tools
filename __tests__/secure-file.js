import { tmpdir } from 'os'
import { join, resolve } from 'path'

import {
  createFileHash,
  decodeBox,
  decryptContents,
  encodeBox,
  encryptContents,
  readEncryptedFile,
  readSecureFile,
  readSignedFile,
  signContents,
  verifyContents,
  writeEncryptedFile,
  writeSecureFile,
  writeSignedFile,
} from '../packages/secure-file'
import { decodeBase64, encodeBase64 } from '../packages/utils-base64'
import {
  PASSWORDHASH_ALG_ARGON2ID13,
  PASSWORDHASH_MEMLIMIT_MIN,
  PASSWORDHASH_OPSLIMIT_MIN,
  createPasswordHashSalt,
  createSecretBoxKey,
  createSecretBoxKeyFromPassword,
  createSignKeyPair,
} from '../packages/utils-crypto'

const getFixture = name => {
  return resolve(__dirname, '__fixtures__/secure-file', name)
}
const getTempFile = name => join(tmpdir(), 'mf-secure-file-test', name)

describe('secure-file', () => {
  describe('encrypt', () => {
    it('provides encodeBox() and decodeBox() to serialize EncryptedBox', () => {
      const encrypted = {
        cipher: Buffer.allocUnsafe(32),
        nonce: Buffer.allocUnsafe(32),
      }
      const encoded = encodeBox(encrypted)
      const decoded = decodeBox(encoded)
      expect(encrypted.cipher.equals(decoded.cipher)).toBe(true)
      expect(encrypted.nonce.equals(decoded.nonce)).toBe(true)
    })

    it('provides encryptContents() and decryptContents()', () => {
      const contents = Buffer.from('contents')
      const key = createSecretBoxKey()

      const encrypted = encryptContents(contents, key)
      expect(encrypted.version).toBe(1)
      expect(encrypted.type).toBe('encrypted')
      expect(encrypted.encrypted).toBeDefined()

      const decrypted = decryptContents(encrypted, key)
      expect(contents.equals(decrypted)).toBe(true)
    })

    it('decryptContents() throws if the version or type is invalid', () => {
      expect(() => decryptContents({ version: 0 })).toThrow('Invalid input')
      expect(() => decryptContents({ version: 1, type: 'something' })).toThrow(
        'Invalid input',
      )
    })

    it('provides writeEncryptedFile() and readEncryptedFile()', async () => {
      const file = getTempFile('encrypt.json')
      const contents = Buffer.from('contents')
      const key = createSecretBoxKey()

      const encrypted = await writeEncryptedFile(file, contents, key)
      const decrypted = await readEncryptedFile(file, key)
      expect(contents.equals(decrypted.opened)).toBe(true)
    })

    it('uses metadata and a key extractor to decrypt the contents', async () => {
      const file = getTempFile('encrypt-meta.json')
      const contents = Buffer.from('contents')
      const password = Buffer.from('not so secure')

      const getKey = meta => {
        return createSecretBoxKeyFromPassword(
          password,
          decodeBase64(meta.salt),
          meta.opslimit,
          meta.memlimit,
          meta.algorithm,
        )
      }
      const metadata = {
        algorithm: PASSWORDHASH_ALG_ARGON2ID13,
        memlimit: PASSWORDHASH_MEMLIMIT_MIN,
        opslimit: PASSWORDHASH_OPSLIMIT_MIN,
        salt: encodeBase64(createPasswordHashSalt()),
      }

      const encrypted = await writeEncryptedFile(
        file,
        contents,
        await getKey(metadata),
        metadata,
      )
      const decrypted = await readEncryptedFile(file, getKey)
      expect(contents.equals(decrypted.opened)).toBe(true)
    })

    it('does not provide the opened buffer if decryption fails', async () => {
      const file = getTempFile('encrypt-fail.json')
      const encrypted = await writeEncryptedFile(
        file,
        Buffer.from('contents'),
        createSecretBoxKey(),
      )
      const notDecrypted = await readEncryptedFile(file, createSecretBoxKey())
      expect(notDecrypted.opened).toBeUndefined()
    })
  })

  describe('fs', () => {
    it('provides readSecureFile() and writeSecureFile()', async () => {
      const file = getTempFile('fs.json')
      await writeSecureFile(file, { test: 'contents' }, { test: 'meta' })
      const data = await readSecureFile(file)
      expect(data).toEqual({
        version: 1,
        contents: { test: 'contents' },
        metadata: { test: 'meta' },
      })
    })

    it('readSecureFile() throws if the file data is invalid', () => {
      const file1 = getFixture('1.not-object.json')
      expect(readSecureFile(file1)).rejects.toThrow('Invalid file')
      const file2 = getFixture('2.no-version.json')
      expect(readSecureFile(file2)).rejects.toThrow('Invalid file')
      const file3 = getFixture('3.bad-version.json')
      expect(readSecureFile(file3)).rejects.toThrow('Invalid file')
      const file4 = getFixture('4.no-contents.json')
      expect(readSecureFile(file1)).rejects.toThrow('Invalid file')
    })
  })

  describe('hash', () => {
    it('createFileHash() returns the hash of the file', async () => {
      const file = getFixture('5.file-hash.txt')
      const hash = await createFileHash(file)
      expect(hash).toMatchSnapshot()
    })
  })

  describe('sign', () => {
    it('provides signContents() and verifyContents()', () => {
      const contents = Buffer.from('contents')
      const kp = createSignKeyPair()

      const signed = signContents(contents, kp)
      expect(signed.version).toBe(1)
      expect(signed.type).toBe('signed')
      expect(signed.signed.keys[0]).toBe(encodeBase64(kp.publicKey))
      expect(signed.signed.message).toBeDefined()

      const verified = verifyContents(signed)
      expect(contents.equals(verified)).toBe(true)
    })

    it('signContents() and verifyContents() handle multiple keys', () => {
      const contents = Buffer.from('contents')
      const keyPairs = [
        createSignKeyPair(),
        createSignKeyPair(),
        createSignKeyPair(),
      ]

      const signed = signContents(contents, keyPairs)
      expect(signed.signed.keys).toHaveLength(3)

      const verified = verifyContents(signed)
      expect(contents.equals(verified)).toBe(true)

      const useKeys = keyPairs.map(kp => kp.publicKey)
      const alsoVerified = verifyContents(signed, useKeys)
      expect(contents.equals(alsoVerified)).toBe(true)

      const failsVerified = verifyContents(signed, useKeys.slice(0, 2))
      expect(failsVerified).toBeUndefined()
    })

    it('signContents() throws an error if an empty array of KeyPairs is provided', () => {
      expect(() => signContents(Buffer.from('test'), [])).toThrow(
        'Missing signing keys',
      )
    })

    it('verifyContents() throws an error if the input is invalid', () => {
      expect(() => verifyContents({})).toThrow('Invalid input')
      expect(() => verifyContents({ version: 0 })).toThrow('Invalid input')
      expect(() => verifyContents({ version: 1, type: 'other' })).toThrow(
        'Invalid input',
      )
    })

    it('provides writeSignedFile() and readSignedFile()', async () => {
      const file = getTempFile('sign.json')
      const contents = Buffer.from('contents')
      const kp = createSignKeyPair()
      const metadata = { test: 'data' }

      await writeSignedFile(file, contents, kp, metadata)
      const data = await readSignedFile(file)

      expect(data.file.version).toBe(1)
      expect(data.file.contents.type).toBe('signed')
      expect(data.file.metadata).toEqual(metadata)
      expect(contents.equals(data.opened)).toBe(true)
    })
  })
})
