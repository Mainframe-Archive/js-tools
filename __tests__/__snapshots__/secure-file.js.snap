// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`secure-file hash createFileHash() returns the hash of the file 1`] = `
Object {
  "data": Array [
    66,
    187,
    169,
    39,
    89,
    87,
    112,
    162,
    210,
    251,
    146,
    61,
    231,
    97,
    174,
    154,
    191,
    86,
    15,
    152,
    16,
    7,
    229,
    105,
    95,
    7,
    14,
    78,
    42,
    22,
    142,
    201,
  ],
  "type": "Buffer",
}
`;
